### ISSUE TITLE
initialDelaySeconds of readinessProbe for kube-dns is too long.

### ISSUE THREAD

**Issue by MrHohn** [kubernetes/kubernetes#33053]
From current [YAML file](https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/skydns-rc.yaml.base#L75) for kube-dns addon.

We are currently using initialDelaySeconds: 30 for readinessProbe for the kubedns container. And [the default state of Readiness before the initial delay is Failure](http://kubernetes.io/docs/user-guide/pod-states/#container-probes). This setup leads to a huge delay between a backend dns server can serve traffic and it is actually ready. Because these service endpoints would not be exposed and iptables rules would not be injected until readiness probe succeed.

I guess 30 seconds initial delay may be make sense when it is the first dns server backend and it need to wait for the Kubernetes master service to be ready. But it is not desired while scaling out kube-dns, in which we would prefer faster reaction time (it takes around 40 seconds for a dns server being exposed in service now).

Would it be feasible to tune down this initalDelay? Even though probing may fail for a while ahead, I guess it wouldn't hurt a lot.

@thockin @girishkalele @bprashanth

**Comment by MrHohn** [comment]
But then failureThreshold would be another concern.

**Comment by thockin** [comment]
I think 5 seconds is reasonable, or even 2 seconds with enough retries.
Try it and run e2e and see how many restarts it hits :)

**Comment by bprashanth** [comment]
can you get ready before you're alive? if not do you need to tune liveness? it can take O(10s) for a list with 100s of pods in the apiserver, and you don't want to crashloop on restart

**Comment by bprashanth** [comment]
don't quote me on the 10s metric, it is stale :)

**Comment by MrHohn** [comment]
> can you get ready before you're alive?

Hmm, I'm not sure about this logic. But as the current definition `The default state of Liveness before the initial delay is Success` says, the liveness probe would always succeed in the initial period. So I guess the  situation a container gets ready before it's alive will not happen.

**Comment by bprashanth** [comment]
so if you set an initial delay for liveness to something really high and an initial delay for readiness to something low, your endpoint gets traffic? that sounds sketchy, because if someone set a high initial delay on liveness it probably means the pod needs that long to boot up (eg: db load from snapshot) and sending it traffic in the interim is going to hit timeouts.

**Comment by bprashanth** [comment]
Anyway it shouldn't *really* matter if you take 5s or 30s to become ready, the deployment should handle it right? you will just have maxSurge dns pods up, and after minReadySeconds, the next pod gets created.

**Comment by MrHohn** [comment]
I think the logic is a bit tricky here.

If I understand correctly, the intention of liveness probe is to make sure the container is still alive, but not to make sure it is up. Instead we should use readiness probe to make sure the container is now able to provide service. And the traffic would be forwarded to endpoints after they are ready. If the container needs that long to boot up, we probably should use high initial delay on readiness probe but not liveness probe.

In summary:
- liveness probe provides infomation for kubelete to restart an unexpected terminated container
- readiness probe provides information for service to know whether the container is ready to go

In this point of view, I instead think the initial delay for liveness probe is suppose to be higher than that of readiness probe.

**Comment by MrHohn** [comment]
I may not understand the comments fully...But I think the whole point here is to well balance these two factors:
- Let those dns backend servers which are ready get exposed in service as soon as possible.
- Make sure those dns backend servers which take a long time to start can take their time without being mistakenly killed.

**Comment by bprashanth** [comment]
Yep. Just try setting it lower than liveness testing it.
